### DISCUSSION TITLE
相似度计算具体过程

### DISCUSSION THREAD

**Discussion by MakoXiao** [milvus-io/milvus#14650] [Q&A and General discussion]
milvus在做相似度计算时，并非进程/线程是从collection中单条与入参向量进行相似度计算，还是小批次(N)做成矩阵和入参向量一起计算呢？ 是下面哪种方式呢？
1. 【1 * 128 维度 】 * 【1 * 128 维度 】   =>  相似度
2. 【1 * 128 维度 】 * 【N * 128 维度 】  =>  N个相似度

**Comment by yanliang567** [comment]
@yhmo  could you please help to answer the question above?

**Reply by yhmo** [reply]
Milvus split data into segments, a search request will build one task for each segment, each task is independent, the search engine executes the tasks one by one. In one task, typically it creates multiple threads according to your machine's CPU cores to calculate similarity batch by batch.
Assume your machine has 8 CPU cores, there are 8 million vectors in a collection and split into 10 segments, each segment 800k vectors. A search request will build 10 tasks to execute. In one task, it creates 8 threads to parallelly compute, each thread will compute for 100k vectors.

**Reply by MakoXiao** [reply]
Thanks for your answer.
forward ，when each thread compute 100k vectors ,  it is one by one calculate vector similarity, or use GPUs to calculate vectors matrix , for batch ?

**Reply by yhmo** [reply]
Actually, the runtime thread mode is more complicated than what I say. The Milvus index engine is based on Faiss, the algorithm is affected by many factors such as CPU instructions, CPU L3 cache size, CPU cores, query vectors number, segment vectors number, vector dimension, topk value, etc. For GPU, the thread model is different. For more information, you can read document from the [Faiss](https://github.com/facebookresearch/faiss) project.

**Reply by MakoXiao** [reply]
Thank expert for your answers
